tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp

```diff
@@ -138,24 +140,29 @@ void EditorDiagConsumer::handleDiagnostic(
     }
   }
 
-  SKInfo.Offset = SM.getLocOffsetInBuffer(Loc, BufferID);
-  std::tie(SKInfo.Line, SKInfo.Column) = SM.getLineAndColumn(Loc, BufferID);
-  SKInfo.Filename = SM.getIdentifierForBuffer(BufferID);
+  if (BufferID) {
+    SKInfo.Offset = SM.getLocOffsetInBuffer(Loc, *BufferID);
+    std::tie(SKInfo.Line, SKInfo.Column) = SM.getLineAndColumn(Loc, *BufferID);
+    SKInfo.Filename = SM.getIdentifierForBuffer(*BufferID);
+  } else {
+    SKInfo.Filename = "<unknown>";
+  }
 
   for (auto R : Info.Ranges) {
-    if (R.isInvalid() || SM.findBufferContainingLoc(R.getStart()) != BufferID)
+    if (R.isInvalid() || !BufferID ||
+        SM.findBufferContainingLoc(R.getStart()) != *BufferID)
       continue;
-    unsigned Offset = SM.getLocOffsetInBuffer(R.getStart(), BufferID);
+    unsigned Offset = SM.getLocOffsetInBuffer(R.getStart(), *BufferID);
     unsigned Length = R.getByteLength();
     SKInfo.Ranges.push_back({ Offset, Length });
   }
 
   for (auto F : Info.FixIts) {
-    if (F.getRange().isInvalid() ||
-        SM.findBufferContainingLoc(F.getRange().getStart()) != BufferID)
+    if (F.getRange().isInvalid() || !BufferID ||
+        SM.findBufferContainingLoc(F.getRange().getStart()) != *BufferID)
       continue;
     unsigned Offset = SM.getLocOffsetInBuffer(F.getRange().getStart(),
-                                              BufferID);
+                                              *BufferID);
     unsigned Length = F.getRange().getByteLength();
     SKInfo.Fixits.push_back({ Offset, Length, F.getText() });
   }
```

Comment:
I'm not sure this makes much difference, so up to you, but I thought the above might be more readable with a top-level `if (BufferId) { check for non-input buffer, set filename + offset + ranges + fixits } else { set unknown filename }`

Reply:
Ah, good point.  I'll do that.